### PR TITLE
Remove tx logs and block bloom from chain state

### LINKED ISSUE #452: Problem: tx logs and block bloom inconsistency in export upgrade
__System info:__ ethermint main branch

__Steps to reproduce:__

Do an export upgrade to ethermint chain, there are several issues:
- firstly of all, block blooms are not exported, and not set in `InitGenesis`.
- even if we fix the block bloom issue, if one reset initial-height, the block numbers referenced in tx logs are invalid.

__Expected behavior:__ should work with export upgrade

__Actual behavior:__ don't work

__Additional info:__ 

Proposal: Don't store tx logs and block bloom in iavl storage, we could emit them in cosmos events, and query from tendermint index.

### PR DESCRIPTION
**Pull request by yihuang**
Closes #452

## Description

- tx logs are stored in transient storage and emitted as events
- block bloom is emitted as event in end block
- RPC API query tx logs block bloom from events of tx result and block result.

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/backend/backend.go [hunk 1]
```diff
@@ -177,6 +178,24 @@ func (e *EVMBackend) GetBlockByHash(hash common.Hash, fullTx bool) (map[string]i
...
+		return ethtypes.Bloom{}, err
+	}
+	for _, event := range result.EndBlockEvents {
+		if event.Type == evmtypes.EventTypeBlockBloom {
+			for _, attr := range event.Attributes {
+				if bytes.Equal(attr.Key, []byte(evmtypes.AttributeKeyEthereumBloom)) {
+					return ethtypes.BytesToBloom(attr.Value), nil
+				}
+			}
+		}
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:194]:
```suggestion
		if event.Type != evmtypes.EventTypeBlockBloom {
			continue
		}
		
		for _, attr := range event.Attributes {
			if bytes.Equal(attr.Key, []byte(evmtypes.AttributeKeyEthereumBloom)) {
				return ethtypes.BytesToBloom(attr.Value), nil
			}
		}
```

File: ethereum/rpc/backend/utils.go [hunk 2]
```diff
@@ -130,3 +133,20 @@ func (e *EVMBackend) getAccountNonce(accAddr common.Address, pending bool, heigh
...
+	for _, event := range events {
+		if event.Type == evmtypes.EventTypeTxLog {
+			for _, attr := range event.Attributes {
+				if bytes.Equal(attr.Key, []byte(evmtypes.AttributeKeyTxLog)) {
+					var log evmtypes.Log
+					codec.MustUnmarshal(attr.Value, &log)
+					logs = append(logs, &log)
+				}
+			}
+		}
```
Comment (fedekunze) [inline ethereum/rpc/backend/utils.go:149]:
```suggestion
		if event.Type != evmtypes.EventTypeTxLog {
			continue
		}
		for _, attr := range event.Attributes {
			if !bytes.Equal(attr.Key, []byte(evmtypes.AttributeKeyTxLog)) {
				continue
			}
			
			var log evmtypes.Log
			codec.MustUnmarshal(attr.Value, &log)
			logs = append(logs, &log)
		}
```

**Review by fedekunze — COMMENTED** [review]
ACK. I would also add documentation on the JSON-RPC to turn on event indexing so that people don't run into the issue of not finding the bloom filter or logs.

This change is state machine braking so we also need to add an entry on the CHANGELOG

**Comment by yihuang** [comment]
> ACK. I would also add documentation on the JSON-RPC to turn on event indexing so that people don't run into the issue of not finding the bloom filter or logs.

In this PR, we actually only used the `/block_results` and `/tx?hash=` tm rpc to get what we want, I think they are available even when tx indexer is not enabled?

> 
> This change is state machine braking so we also need to add an entry on the CHANGELOG

changelog added.

**Comment by fedekunze** [comment]
Thanks @yihuang feel free to fix the linter with `make lint-fix`

**Review by fedekunze — APPROVED** [review]

**Pull request merged**